FIPS related docs

mposolda opened the pull request:
Closes #16444 #12432 #12429

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

mposolda committed:
FIPS related docs
Closes #16444 #12432 #12429

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
(cherry picked from commit 62c706fd43ec4b9aef07f7c91bc4f3f0f43a97ae)

mposolda commented:
@pedroigor @rmartinc @andymunro Do you please have opportunity to review this PR for FIPS documentation?

There will be still some slight changes to it due the work #17208 , #17210 and #17211, but overally it should be ready for the review.

andymunro commented on docs/fips.md:
```diff
@@ -1,171 +1,9 @@
-
-Run the integration tests in the FIPS environment
--------------------------------------------------
+Run the unit or integration tests in the FIPS environment
```
```suggestion
Run the unit or integration tests in the FIPS environment.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+In order to run in FIPS 140-2, Keycloak should run on FIPS 140-2 enabled system. This usually assumes RHEL or Fedora where FIPS was enabled during installation.
+See https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html-single/security_hardening/index#assembly_installing-the-system-in-fips-mode_security-hardening[RHEL documentation]
+for the details. When system is in FIPS mode, it will automatically make sure that underlying OpenJDK is in FIPS mode as well and would use only
```
```suggestion
for the details. When the system is in FIPS mode, it makes sure that the underlying OpenJDK is in FIPS mode as well and would use only
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+== BouncyCastle library
+
+Keycloak internally uses BouncyCastle library for lots of cryptography utilities. The thing is, that the default flavour of the BouncyCastle library shipped with Keycloak is not FIPS compliant.
```
```suggestion
Keycloak internally uses the BouncyCastle library for many cryptography utilities. However, the default flavor of the BouncyCastle library that shipped with Keycloak is not FIPS compliant.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+Keycloak internally uses BouncyCastle library for lots of cryptography utilities. The thing is, that the default flavour of the BouncyCastle library shipped with Keycloak is not FIPS compliant.
+However BouncyCastle also provides FIPS validated version of it's library. The FIPS validated BouncyCastle library cannot be shipped with Keycloak by default for the licence purpose and
+Keycloak cannot provide official support of it. So the requirement is, that to run in FIPS compliant mode, you need to download BouncyCastle-FIPS bits and add them to the Keycloak distribution.
```
```suggestion
Keycloak cannot provide official support of it. Therefore, to run in FIPS compliant mode, you need to download BouncyCastle-FIPS bits and add them to the Keycloak distribution.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+=== BouncyCastle FIPS bits
+
+BouncyCastle FIPS can be downloaded from https://www.bouncycastle.org/fips-java/[BouncyCastle official page]. Then you can add them to the directory
```
```suggestion
BouncyCastle FIPS can be downloaded from the https://www.bouncycastle.org/fips-java/[BouncyCastle official page]. Then you can add them to the directory
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+== Generating keystore
+
+You can create either `pkcs12` or `bcfks` keystore to be used for Keycloak server SSL. The `pkcs12` works fine just in BCFIPS non-approved mode.
```
```suggestion
You can create either `pkcs12` or `bcfks` keystore to be used for the Keycloak server SSL. The `pkcs12` works well in BCFIPS non-approved mode.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+PKCS12 keystore can be generated with OpenJDK 17 Java on RHEL 9 in the standard way.
+
+For the BCFKS keystore generation, it is needed to use the BouncyCastle FIPS libraries and use custom security file.
```
```suggestion
BCFKS keystore generation requires the use of the BouncyCastle FIPS libraries and a custom security file.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+For the BCFKS keystore generation, it is needed to use the BouncyCastle FIPS libraries and use custom security file.
+
+You can start with create some helper file, for instance `/tmp/kc.keystore-create.java.security`. The content of the file just needs to have this single property
```
```suggestion
You can start with create a helper file, such as `/tmp/kc.keystore-create.java.security`. The content of the file needs only to have this single property
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
````diff
@@ -0,0 +1,229 @@
+securerandom.strongAlgorithms=PKCS11:SunPKCS11-NSS-FIPS
+```
+
+Then use the command like this to generate the keystore:
````
```suggestion
Next enter a command such as the following to generate the keystore:
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
````diff
@@ -0,0 +1,229 @@
+  -J-Djava.security.properties=/tmp/kc.keystore-create.java.security
+```
+
+WARNING: Using self-signed certificates is for demonstration purposes only, please make sure to use proper certificates in production instead
````
```suggestion
WARNING: Using self-signed certificates is for demonstration purposes only, so replace these cerrtificates with proper certificates when you move to a production environment.
```

mposolda replied:
@andymunro you made a small typo in the word `cerrtificates`. But I will fix that in the follow-up commit...

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+== Running the server.
+
+To run the server with BCFIPS in non-approved mode:
```
```suggestion
To run the server with BCFIPS in non-approved mode, enter the following command::
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+<@kc.start parameters="--fips-mode=enabled --hostname=localhost --https-key-store-password=passwordpassword --log-level=INFO,org.keycloak.common.crypto:TRACE,org.keycloak.crypto:TRACE"/>
+
+NOTE: The logging can be disabled in production if everything works as expected.
```
```suggestion
NOTE: You can disable logging in production if everything works as expected.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+== Strict mode
+
+For the `fips-mode` option, the more secure alternative is to use `--fips-mode=strict` in which case BouncyCastle FIPS will use "approved mode".
+That means even stricter security requirements on cryptography and security algorithms.
```
```suggestion
Using that option results in stricter security requirements on cryptography and security algorithms.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+For the `fips-mode` option, the more secure alternative is to use `--fips-mode=strict` in which case BouncyCastle FIPS will use "approved mode".
+That means even stricter security requirements on cryptography and security algorithms.
+
+When starting server at startup, you can check that startup log contains `KC` provider contains KC provider with the note about `Approved Mode` like this:
```
```suggestion
When starting server, you can check that the startup log contains `KC` provider with the note about `Approved Mode` such as the following:
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+=== Cryptography restrictions in strict mode
+
+* As mentioned above, strict mode may not work with `pkcs12` keystore. It is needed to use other keystore (like `bcfks`) as mentioned above. Also `jks` and `pkcs12` keystores are not
```
```suggestion
* As mentioned in the previous section, strict mode may not work with `pkcs12` keystore. It is needed to use another keystore (like `bcfks`) as mentioned earlier. Also `jks` and `pkcs12` keystores are not
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+=== Cryptography restrictions in strict mode
+
+* As mentioned above, strict mode may not work with `pkcs12` keystore. It is needed to use other keystore (like `bcfks`) as mentioned above. Also `jks` and `pkcs12` keystores are not
+supported in Keycloak itself. For instance when importing/generating keystore of OIDC/SAML client in the admin console or for `java-keystore` provider in the realm keys.
```
```suggestion
supported in Keycloak. Some examples are importing or generating a keystore of an OIDC or SAML client in the Admin Console or for a `java-keystore` provider in the realm keys.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+supported in Keycloak itself. For instance when importing/generating keystore of OIDC/SAML client in the admin console or for `java-keystore` provider in the realm keys.
+
+* User passwords must be 14 characters or longer. Keycloak uses PBKDF2 based password encoding by default. BCFIPS approved mode requires passwords to be at least 112 bits
+(effectively 14 characters) with PBKDF2 algorithm. If you want to allow shorter password, you need to set property `max-padding-length` of provider `pbkdf2-sha256` of SPI `password-hashing`
```
```suggestion
(effectively 14 characters) with PBKDF2 algorithm. If you want to allow a shorter password, set the property `max-padding-length` of provider `pbkdf2-sha256` of SPI `password-hashing`
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+* User passwords must be 14 characters or longer. Keycloak uses PBKDF2 based password encoding by default. BCFIPS approved mode requires passwords to be at least 112 bits
+(effectively 14 characters) with PBKDF2 algorithm. If you want to allow shorter password, you need to set property `max-padding-length` of provider `pbkdf2-sha256` of SPI `password-hashing`
+to value 14, so there will be some additional padding used when verifying hash created by this algorithm. This is also backwards compatible with previously stored passwords
```
```suggestion
to value 14 to provide additional padding when verifying a hash created by this algorithm. This setting is also backwards compatible with previously stored passwords
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+* User passwords must be 14 characters or longer. Keycloak uses PBKDF2 based password encoding by default. BCFIPS approved mode requires passwords to be at least 112 bits
+(effectively 14 characters) with PBKDF2 algorithm. If you want to allow shorter password, you need to set property `max-padding-length` of provider `pbkdf2-sha256` of SPI `password-hashing`
+to value 14, so there will be some additional padding used when verifying hash created by this algorithm. This is also backwards compatible with previously stored passwords
+(if you had your user's DB in non-FIPS environment and you have shorter passwords and you want to verify them now with Keycloak using BCFIPS in approved mode, it should work fine).
```
```suggestion
For example, if the user's database is in a non-FIPS environment and you have shorter passwords and you want to verify them now with Keycloak using BCFIPS in approved mode, the passwords should work.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+(effectively 14 characters) with PBKDF2 algorithm. If you want to allow shorter password, you need to set property `max-padding-length` of provider `pbkdf2-sha256` of SPI `password-hashing`
+to value 14, so there will be some additional padding used when verifying hash created by this algorithm. This is also backwards compatible with previously stored passwords
+(if you had your user's DB in non-FIPS environment and you have shorter passwords and you want to verify them now with Keycloak using BCFIPS in approved mode, it should work fine).
+So effectively, you can use option like this when starting the server:
```
```suggestion
So effectively, you can use an option such as the following when starting the server:
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
````diff
@@ -0,0 +1,229 @@
+--spi-password-hashing-pbkdf2-sha256-max-padding-length=14
+```
+
+NOTE: Using the option above does not break FIPS compliance. However note that longer passwords are good practice anyway. For example passwords auto-generated by modern browsers match this
````
```suggestion
NOTE: Using the option above does not break FIPS compliance. However, note that longer passwords are good practice anyway. For example. passwords auto-generated by modern browsers match this
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+NOTE: Using the option above does not break FIPS compliance. However note that longer passwords are good practice anyway. For example passwords auto-generated by modern browsers match this
+requirement as they are longer than 14 characters.
+
+* RSA keys of 1024 bits do not work (2048 is the minimum). This applies for keys used by Keycloak realm itself (Realm keys from the `Keys` tab in the admin console), but also client keys and IDP keys
```
```suggestion
* RSA keys of 1024 bits do not work (2048 is the minimum). This applies for keys used by the Keycloak realm itself (Realm keys from the `Keys` tab in the admin console), but also client keys and IDP keys
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+== Other restrictions
+
+In order to have SAML working, there is a need to have `XMLDSig` security provider to be available in your security providers.
```
```suggestion
To have SAML working, make sure that a`XMLDSig` security provider is available in your security providers.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+== Other restrictions
+
+In order to have SAML working, there is a need to have `XMLDSig` security provider to be available in your security providers.
+In order to have Kerberos working, there is a need to have `SunJGSS` security provider available. In FIPS enabled RHEL 9 in OpenJDK 17.0.6, these
```
```suggestion
To have Kerberos working, make sure that a `SunJGSS` security provider is available. In FIPS enabled RHEL 9 in OpenJDK 17.0.6, these
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+In order to have Kerberos working, there is a need to have `SunJGSS` security provider available. In FIPS enabled RHEL 9 in OpenJDK 17.0.6, these
+security providers are not by default in the `java.security`, which means that they effectively cannot work.
+
+To have SAML working, you can manually add the provider into `JAVA_HOME/conf/security/java.security` into the list fips providers. For example add the line like:
```
```suggestion
To have SAML working, you can manually add the provider into `JAVA_HOME/conf/security/java.security` into the list fips providers. For example add the line such as the following:
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
````diff
@@ -0,0 +1,229 @@
+fips.provider.7=XMLDSig
+```
+
+Adding this security provider should be fine as in fact it is FIPS compliant and likely will be added by default in the future OpenJDK 17 micro version.
````
```suggestion
Adding this security provider should work well. In fact, it is FIPS compliant and likely will be added by default in the future OpenJDK 17 micro version.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+== Run the CLI on the FIPS host
+
+In case you want to run Client Registration CLI (`kcreg.sh|bat` script) or Admin CLI (`kcadm.sh|bat` script), it is needed
```
```suggestion
If you want to run Client Registration CLI (`kcreg.sh|bat` script) or Admin CLI (`kcadm.sh|bat` script), 
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+== Run the CLI on the FIPS host
+
+In case you want to run Client Registration CLI (`kcreg.sh|bat` script) or Admin CLI (`kcadm.sh|bat` script), it is needed
+that CLI will also use the BouncyCastle FIPS dependencies instead of plain BouncyCastle dependencies. To achieve this, you may copy the
```
```suggestion
 the CLI must also use the BouncyCastle FIPS dependencies instead of plain BouncyCastle dependencies. To achieve this, you may copy the
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+In case you want to run Client Registration CLI (`kcreg.sh|bat` script) or Admin CLI (`kcadm.sh|bat` script), it is needed
+that CLI will also use the BouncyCastle FIPS dependencies instead of plain BouncyCastle dependencies. To achieve this, you may copy the
+jars to the CLI library folder and that is enough. CLI tool will automatically use BCFIPS dependencies instead of plain BC when
+it detects that corresponding BCFIPS jars are present (see above for the versions used). For example use command like this before running CLI:
```
```suggestion
it detects that corresponding BCFIPS jars are present (see above for the versions used). For example, use command like this before running CLI:
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+== Keycloak server in FIPS mode in the container
+
+When you want Keycloak in FIPS mode to be executed inside container, it is needed that your "host" is using FIPS mode as well. The container
```
```suggestion
When you want Keycloak in FIPS mode to be executed inside a container, your "host" is must be using FIPS mode as well. The container
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+See https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/security_hardening/using-the-system-wide-cryptographic-policies_security-hardening#enabling-fips-mode-in-a-container_using-the-system-wide-cryptographic-policies[this section]
+in the RHEL documentation for the details.
+
+So keycloak container image will be just automatically in fips mode when executed from the host in FIPS mode. So only things needed is to
```
```suggestion
The keycloak container image will be just automatically in fips mode when executed from the host in FIPS mode. 
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+in the RHEL documentation for the details.
+
+So keycloak container image will be just automatically in fips mode when executed from the host in FIPS mode. So only things needed is to
+make sure that Keycloak container also uses BCFIPS jars (instead of BC jars) and proper options when started.
```
```suggestion
However, make sure that the Keycloak container also uses BCFIPS jars (instead of BC jars) and proper options when started.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+ENTRYPOINT ["/opt/keycloak/bin/kc.sh"]
+----
+
+Then build FIPS optimized docker image and start it as described in the <@links.server id="containers"/>. Assumption is to use arguments as described above when starting the image.
```
```suggestion
Then build FIPS an optimized docker image and start it as described in the <@links.server id="containers"/>. These steps require that you use arguments as described above when starting the image.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+== Migration from non-fips environment
+
+If you previously used Keycloak in the non-fips environment, it is possible to migrate it to FIPS environment including it's data. However there are restrictions and considerations as
```
```suggestion
If you previously used Keycloak in the non-fips environment, it is possible to migrate it to FIPS environment including it's data. However, restrictions and considerations exist as
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+== Migration from non-fips environment
+
+If you previously used Keycloak in the non-fips environment, it is possible to migrate it to FIPS environment including it's data. However there are restrictions and considerations as
+mentioned above. To highlight some of them:
```
```suggestion
mentioned in previous sections. To highlight some of them:
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+* Kerberos authentication may not work. If your authentication flow uses `Kerberos` authenticator, this authenticator will be automatically switched to `DISABLED` when migrated to FIPS
+environment. It is recommended to remove any `Kerberos` user storage providers from your realm and disable `Kerberos` related functionality in LDAP providers before switching to FIPS environment.
+
+In addition to the above, it is recommended to doublecheck this before switching to FIPS strict mode:
```
```suggestion
In addition to the preceding requirements, be sure to doublecheck this before switching to FIPS strict mode:
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+In addition to the above, it is recommended to doublecheck this before switching to FIPS strict mode:
+
+* Make sure that all the Keycloak functionality relying on keys (EG. realm or client keys) use RSA keys of at least 2048 bits
```
```suggestion
* Make sure that all the Keycloak functionality relying on keys (for example, realm or client keys) use RSA keys of at least 2048 bits
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+* Make sure that clients relying on `Signed JWT with Client Secret` use at least 14 characters long secrets (ideally generated secrets)
+
+* Password length restriction as described above. In case your users have shorter passwords, it is needed to start the server with the max padding length set to 14 of PBKDF2 provider as mentioned
```
```suggestion
* Password length restriction as described earlier. In case your users have shorter passwords, be sure to start the server with the max padding length set to 14 of PBKDF2 provider as mentioned
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+* Make sure that clients relying on `Signed JWT with Client Secret` use at least 14 characters long secrets (ideally generated secrets)
+
+* Password length restriction as described above. In case your users have shorter passwords, it is needed to start the server with the max padding length set to 14 of PBKDF2 provider as mentioned
+above. If you prefer to avoid this option, you can for instance ask all your users to reset their password (for example by `Forget password` link) during first authentication in the new environment.
```
```suggestion
earlier. If you prefer to avoid this option, you can for instance ask all your users to reset their password (for example by the `Forget password` link) during the first authentication in the new environment.
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+== Keycloak FIPS mode on the non-fips system
+
+Keycloak is tested on FIPS enabled RHEL 9 system and `ubi9` image. Running on the non-RHEL compatible platform or on the non-FIPS enabled platform, the FIPS compliance cannot be
```
```suggestion
Keycloak is tested on a FIPS enabled RHEL 9 system and `ubi9` image. Running on the non-RHEL compatible platform or on the non-FIPS enabled platform, the FIPS compliance cannot be
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+Keycloak is tested on FIPS enabled RHEL 9 system and `ubi9` image. Running on the non-RHEL compatible platform or on the non-FIPS enabled platform, the FIPS compliance cannot be
+strictly guaranteed and cannot be officially supported.
+
+If you are still restricted to run Keycloak on such system, you can for instance at least update your security providers configured in `java.security` file. This does not mean FIPS comliance, but
```
```suggestion
If you are still restricted to run Keycloak on such a system, you can at least update your security providers configured in `java.security` file. This update does not mean FIPS compliance, but
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+strictly guaranteed and cannot be officially supported.
+
+If you are still restricted to run Keycloak on such system, you can for instance at least update your security providers configured in `java.security` file. This does not mean FIPS comliance, but
+at least the setup is closer to it. It can be done by providing custom security file with only overriden list of security providers as described above. The list of recommended providers can be
```
```suggestion
at least the setup is closer to it. It can be done by providing a custom security file with only an overriden list of security providers as described earlier. For a list of recommended providers, 
```

andymunro replied:
```suggestion
at least the setup is closer to it. It can be done by providing custom security file with only overriden list of security providers as described above. The list of recommended providers can be
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+If you are still restricted to run Keycloak on such system, you can for instance at least update your security providers configured in `java.security` file. This does not mean FIPS comliance, but
+at least the setup is closer to it. It can be done by providing custom security file with only overriden list of security providers as described above. The list of recommended providers can be
+seen from the https://access.redhat.com/documentation/en-us/openjdk/17/html/configuring_openjdk_17_on_rhel_with_fips/openjdk-default-fips-configuration[OpenJDK 17 documentation].
```
```suggestion
see the https://access.redhat.com/documentation/en-us/openjdk/17/html/configuring_openjdk_17_on_rhel_with_fips/openjdk-default-fips-configuration[OpenJDK 17 documentation].
```

andymunro commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+at least the setup is closer to it. It can be done by providing custom security file with only overriden list of security providers as described above. The list of recommended providers can be
+seen from the https://access.redhat.com/documentation/en-us/openjdk/17/html/configuring_openjdk_17_on_rhel_with_fips/openjdk-default-fips-configuration[OpenJDK 17 documentation].
+
+You can check Keycloak server log at startup if correct security providers are used (TRACE logging should be enabled for crypto related Keycloak packages as described in the Keycloak startup command above).
```
```suggestion
You can check the Keycloak server log at startup to see if the correct security providers are used. TRACE logging should be enabled for crypto-related Keycloak packages as described in the Keycloak startup command earlier.
```

andymunro commented on docs/guides/src/main/server/keycloak-truststore.adoc:
```diff
@@ -39,6 +39,10 @@ For HTTPS requests, this option verifies the hostname of the server's certificat
 * When using `STRICT`, the Common Name (CN) must match the hostname exactly.
 
+type::
+The type of truststore like for example `jks`, `pkcs12` or `bcfks`. If not provided, the type would be detected based on the truststore
```
```suggestion
The type of truststore, such as `jks`, `pkcs12` or `bcfks`. If not provided, the type would be detected based on the truststore
```

andymunro commented on testsuite/integration-arquillian/HOW-TO-RUN.md:
```diff
@@ -948,8 +948,22 @@ networks dns configuration and run the tests.
+
+### Integration tests
+
+On the FIPS enabled platform with FIPS enabled OpenJDK 17, you can run this to test against Keycloak server on Quarkus
```
```suggestion
On the FIPS enabled platform with FIPS enabled OpenJDK 17, you can run this to test against a Keycloak server on Quarkus
```

andymunro reviewed (changes requested):
Thanks for asking for input @mposolda.  A few suggestions for you.

mposolda commented:
@andymunro Thanks a lot for the deep review!

I've applied all your suggestions. I've noticed some minor typos in your suggestions, so added an additional commit to this PR to fix them: https://github.com/keycloak/keycloak/pull/17196/commits/53748e0925d1ef7af74888b938a1c07b20423b58 . You can check if the changes makes sense to you. If not, feel free to propose another suggestion to improve them. Thanks

andymunro reviewed (dismissed):
Looks good, @mposolda.  Thanks for the updates.

rmartinc commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+To run the server with BCFIPS in non-approved mode, enter the following command::
+
+<@kc.start parameters="--fips-mode=enabled --hostname=localhost --https-key-store-password=passwordpassword --log-level=INFO,org.keycloak.common.crypto:TRACE,org.keycloak.crypto:TRACE"/>
```
Now it can be `--features=fips` only and it defaults to `--fips-mode=non-strict` although the latter option can also be passed with `non-strict` or `strict` values.

rmartinc commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+=== Cryptography restrictions in strict mode
+
+* As mentioned in the previous section, strict mode may not work with `pkcs12` keystore. It is needed to use another keystore (like `bcfks`) as mentioned earlier. Also `jks` and `pkcs12` keystores are not
+supported in Keycloak. Some examples are importing or generating a keystore of an OIDC or SAML client in the Admin Console or for a `java-keystore` provider in the realm keys.
```
Why pkcs12 is not supported in keyloak, you mean in strict mode? Right? If I'm right I would add `Also jks and pkcs12 keystores are not supported in Keycloak when using strict mode`.

rmartinc commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+== Other restrictions
+
+To have SAML working, make sure that a`XMLDSig` security provider is available in your security providers.
```
Nitpicking but separate "a `XMLSig`".

rmartinc commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+RUN cp /tmp/files/keycloak-fips.keystore.* /opt/keycloak/conf/server.keystore
+RUN cp /tmp/files/kc.java.security /opt/keycloak/conf/
+
+RUN /opt/keycloak/bin/kc.sh build --fips-mode=strict
```
--features=fips --fips-mode=strict

rmartinc commented on docs/guides/src/main/server/fips.adoc:
```diff
@@ -0,0 +1,229 @@
+
+== Strict mode
+
+For the `fips-mode` option, the more secure alternative is to use `--fips-mode=strict` in which case BouncyCastle FIPS will use "approved mode".
```
to use `--features=fips --fips-mode=strict`

rmartinc reviewed (changes requested):
@mposolda Comments about the new start options and the fips preview feature added.

mposolda committed:
Docs review updates

mposolda commented:
@rmartinc I've added your suggestions. I've used separate commit for the easier review if you want to doublecheck just the latest changes: https://github.com/keycloak/keycloak/pull/17196/commits/72a17d78164d5c70e9a7a017b97bbd863254c33d

I plan to squash everything before merge. Also added note about FIPS being "preview" feature.

Can you please re-review?
